### ISSUE TITLE
second 'non-empty patterns' error silenced by async function

### ISSUE THREAD

**Issue by clubby789** [rust-lang/rust#111537]
I tried this code:

```rust
pub fn bar() {
    match true {};
}

pub async fn foo() {
    match false {};
}
```

I expected to see this happen: Two `error[E0004]: non-exhaustive patterns: type `bool` is non-empty`

Instead, this happened: Only one is reported, in the async function. The error in the async function is still reported if the order is flipped. Making both functions sync or async correctly reports both errors. If there is any code following the faulty match, it is always reported as unreachable, despite no error being emitted for the preceding line in the sync function.

Found in #111533 test

### Meta
`rustc --version --verbose`:
```
rustc 1.71.0-nightly (4a59ba4d5 2023-05-12)
binary: rustc
commit-hash: 4a59ba4d54a3ec0d8ea1e82b7eeb5c8b0162de04
commit-date: 2023-05-12
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.2
```

**Comment by clubby789** [comment]
Actually this seems a bit more general.
```rs
pub fn bar() {
    usize::MAX + 1;
    return;
}

pub async fn foo() {
    usize::MAX + 1;
    return;
}
```
only fires one error (in the sync function this time), but two warnings (for unused arithmetic)

This also happens if the errors are slightly different:
```rs
pub fn bar() {
    match true {};
    return;
}

pub async fn foo() {
    usize::MAX + 1;
    return;
}
```

```
warning: unreachable statement
 --> poc.rs:3:5
  |
2 |     match true {};
  |     ------------- any code following this expression is unreachable
3 |     return;
  |     ^^^^^^^ unreachable statement
  |
  = note: `#[warn(unreachable_code)]` on by default

error[E0004]: non-exhaustive patterns: type `bool` is non-empty
 --> poc.rs:2:11
  |
2 |     match true {};
  |           ^^^^
  |
  = note: the matched value is of type `bool`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern as shown
  |
2 ~     match true {
3 +         _ => todo!(),
4 ~     };
  |
```
The match error now only in the sync function, and the unused arithmetic warning is missing
